### ISSUE TITLE
Chart within a Window control

### ISSUE THREAD

**Issue by sforsherman** [telerik/kendo-angular#185]
Hi guys, I am helping a prospect / customer to raise this issue because they have high security constrains and are unable to share the main application coding online or externally.

During the time that they have been using the Chart control, they have discovered that when the Chart is placed within the page itself, there is no issues. But when they place the Chart in a Window control, it doesn't seem to render.

After some looking through of the chart.component.js file, we discovered that it had something to do with the Chart decorators:

`ChartComponent.decorators = [
        { type: core_1.Component, args: [{
                    exportAs: 'kendoChart',
                    providers: [configuration_service_1.ConfigurationService, tooltip_template_service_1.TooltipTemplateService],
                    selector: 'kendo-chart',
                    template: "\n    <div class=\"k-chart-surface\" (mouseleave)=\"chartMouseleave($event)\"></div>\n    <kendo-chart-crosshair-tooltips-container>\n    </kendo-chart-crosshair-tooltips-container>\n    <kendo-chart-tooltip-popup (mouseleave)=\"tooltipMouseleave($event)\">\n    </kendo-chart-tooltip-popup>\n    <kendo-resize-sensor (resize)=\"chartResize()\"></kendo-resize-sensor>\n  "
                },] },
    ];`

And because the decorators are not added to the chart surface within the Window control, the chart can't function properly.

So in an effort to test out the theory, we created a simple hack to test it, and this was what the customer did on their end,

`ChartComponent.prototype.init = function () {
        var _this = this;
        if (typeof document === 'undefined') {
            return;
        }
        var element = this.element.nativeElement.querySelector('.k-chart-surface');
        if (!element) {
            // ================================================
            // added hack here
            var div = document.createElement('div')
            div.classList.add('k-chart-surface')
            var chart = this.element.nativeElement.querySelector('.k-chart')
            this.element.nativeElement.removeChild(chart)
            this.element.nativeElement.appendChild(div)
            element = this.element.nativeElement.querySelector('.k-chart-surface');
            // =================================================
            // return;
        }
        this.ngZone.runOutsideAngular(function () {
            _this.options.render = function () { _this.render.emit(); };
            _this.instance = new charts.Chart(element, _this.options, _this.theme, _this.intl);
            _this.bindEvents();
        });
    };`

The prospect is looking for a solution soon as this is going to affect their project's progress.

**Comment by tsvetomir** [comment]
That's very interesting indeed and we'd love to address this issue. That said, I have no idea how to replicate the issue. Can you help me with a plunkr or a sample repo that I can test with?

**Comment by giftofjehovah** [comment]
Hi @tsvetomir, unfortunately we are not able to share our code with you. I suggest  we can do a screen share and I can share with you what we are experiencing.

**Comment by tsvetomir** [comment]
We offer remote assistance only as part of our DevCraft Ultimate bundle. You can submit a support ticket in our system and continue there.

Can you please clarify what is a "Window control" in the context of your app?

**Comment by giftofjehovah** [comment]
Hi, I submitted a support ticket and I was redirect to the github repo by your support staff.

Ignore the window part. It just not rendering within another component. How could we solve this problem together? I just need to show you the bug as its quite hard to replicate the issue.

**Comment by tsvetomir** [comment]
Can you please try to modify [my sample](http://plnkr.co/edit/wydKIc77dlMxS0hrPXha?p=preview) to replicate the issue?

Apologies for bouncing you around, but we didn't provide commercial support until yesterday.
Feel free to continue the discussion in a ticket, especially if you want to share code in private.

**Comment by giftofjehovah** [comment]
Hi, 

I think its hard for me to replicate the issue here as its only happening within our application. Let me give a bit more context in summary.

We have 2 applications:

1.    Kitchen Sink – Containing all our components where we extend kendo UI 2 components
2.    Main Application – Containing business logic and using components from kitchen sink
 
We have a particular issue where the Kendo Chart is as expected in Kitchen Sink. However, it is not rendering as expected in our main application. We have tried to debug ourselves and realize the following:

1.    This particular div with a class name of `k-chart-surface` which I believe where the chart is rendered on appeared in our Kitchen Sink.
2.    In our main application, the div with the class name of `k-chart-surface` disappear or does not render, not sure why. As such the chart does not render.
 
I believe it have to be something within our main application that is causing this issue, but we are not able to rectify that.

Under what conditions will the `k-chart-surface` div not render?
Do you have any idea why the chart is behaving this way?

**Comment by tsvetomir** [comment]
As you've noted the `.k-chart-surface` element is rendered as part of the template:

```
  template: `
    <div class="k-chart-surface" (mouseleave)="chartMouseleave($event)"></div>
    ...
```

There's nothing unusual about it as far as I can see.

You mention that _"the decorators are not added to the chart surface within the Window control"_. I'm not sure when that would happen in Angular.

**Comment by giftofjehovah** [comment]
I would appreciate that we can have a screen sharing session. It will be clearer if I can show you the issues I'm facing. 

The entire template of the k-chart is not rendering in my case.

**Comment by dimodi** [comment]
Hello @giftofjehovah ,

My name is Dimo Dimov and I am the support lead at Kendo UI.

Indeed, based on the provided information, and the fact that the unexpected behavior is exhibited in a specific application only, we agree with you on the assumption that the issue is likely to be related to the application's setup or custom logic, and not to the Kendo UI components. From this point of view, investigating the cause of the problem steps into the [scope of our Professional Services team](https://www.telerik.com/account/support/scope). Nevertheless, as this is not 100% certain, we will readily review an isolated sample and suggest next steps - you can provide such a sample in ticket 1083440, which has been opened on this topic, as we now provide official support for Kendo UI for Angular 2 (after we released a go-live RC).

I confirm that realtime [remote web assistance](https://www.telerik.com/account/remote-web-assistance) is available to customers with a DevCraft Ultimate license. 

Regards,
Dimo Dimov

**Comment by giftofjehovah** [comment]
Hi @dimodi ,

How do I provide an isolated sample? I can't send the whole code repo to you due to confidentiality issues. Providing part of the code might not replicate the issue as well.

The only way I think might be the best is to do a screenshare unless there is a better way.
Can I suggest we use join.me?
Its not feasible for me to use DevCraft Ultimate license right and pay $2,199  just for this one session.
Hope you understand and find this request within reason.

**Comment by dimodi** [comment]
Hi @giftofjehovah,

Here are a few quick tips for isolating a given behavior in an isolated sample application:

- remove any irrelevant or sensitive parts from the application, including whole pages, custom HTML, CSS, images, components and scripts. This will conceal the custom design and content, and make the example more focused
- if data is needed in the scenario, replace the actual data that may be sensitive with some (hard-coded) dummy data
- if navigation or complex user interaction is not necessary, include only a single web page in the isolated sample

In short, there should be a way to mimic the implementation of the actual application and the observed issue, but at the same time, remove any sensitive data, imagery or custom irrelevant logic.

Additional thoughts on the matter are provided in the following blog post:

http://www.telerik.com/blogs/isolating-a-problem-in-a-sample-project

My suggestion is to continue the communication in the ticket (1083440), where I have provided this same message as well.

Regards,
Dimo
